eventbus/jetstream/sensor/trigger_conn.go

```diff
@@ -257,15 +257,35 @@ func (conn *JetstreamTriggerConn) processMsg(
 		conn.Logger.Errorf("can't get Metadata() for message %+v??", m)
 	}
 
-	defer func() {
-		err = m.AckSync() // this call does the acknowledgment and waits for an acknowledgment from the server (that it received the ack) too
-		if err != nil {
-			errStr := fmt.Sprintf("Error performing AckSync() on message: %v", err)
-			conn.Logger.Error(errStr)
+	done := make(chan bool)
+	go func() {
+		ticker := time.NewTicker(500 * time.Millisecond)
+		for {
+			select {
+			case <-done:
+				err = m.AckSync()
+				if err != nil {
+					errStr := fmt.Sprintf("Error performing AckSync() on message: %v", err)
+					conn.Logger.Error(errStr)
+				}
+				conn.Logger.Debugf("acked message of Stream seq: %s:%d, Consumer seq: %s:%d", meta.Stream, meta.Sequence.Stream, meta.Consumer, meta.Sequence.Consumer)
+				ticker.Stop()
+				break
+			case <-ticker.C:
+				err = m.InProgress()
+				if err != nil {
+					errStr := fmt.Sprintf("Error performing InProgess() on message: %v", err)
+					conn.Logger.Error(errStr)
+				}
+				conn.Logger.Debugf("InProgess message of Stream seq: %s:%d, Consumer seq: %s:%d", meta.Stream, meta.Sequence.Stream, meta.Consumer, meta.Sequence.Consumer)
+			}
 		}
+	}()
 
-		conn.Logger.Debugf("acked message of Stream seq: %s:%d, Consumer seq: %s:%d", meta.Stream, meta.Sequence.Stream, meta.Consumer, meta.Sequence.Consumer)
+	defer func() {
```

Comment:
This means, no matter the triggering is successful not not, the message will be acked anyways, this is not `atLeastOnce`.

Reply:
When the retry loop is exhausted at some point we have no other choice than to ack the message and move on. This has no bearing on at least once vs at most once semantics.

At most once: attempt invocation only once, if a response is not received (eg: service unavailable) do not attempt again, as we cannot say if the downstream system will eventually succeed.

At least once: attempt invocation multiple times, this may result in the action being invoked multiple times. However, we still need to give up and move on at some point.

Reply:
Like discussed in #1977, we've looked to change the behavior in the case of a SIGFAULT, that's still something we can control to make it from AtMostOnce to AtLeastOnce.

Reply:
Trying to understand what you proposed.

Let's assume there are 1M stacked messages, with async triggering, it's possible that lots of triggering won't be able to get a chance to execute because the goroutine resources might have exhausted, but the messages have been acked, so there's no chance to retrigger. But with sync triggering, this will not happen. Am I right?

Reply:
Yes. The messages will start processing but won’t get acked until the trigger processing completes. 
And we show in the functional test that if a sensor goes down while the message is being processed, it will get processed again.

Reply:
Just to expand a bit on this. When a trigger is satisfied and needs to be invoked we can either:
1. Call the action function _before_ we acknowledge the message (current implementation, albeit in a roundabout way because of the go routine)
2. Call the action function _after_ we acknowledge the message

In both scenarios, our application may crash in between message acknowledgement and calling the action function. But in the second scenario our application will re-process the message on restart. In total there are four possibilities:

| ack | crash | invocations |
| --- | --- | --- |
| before | no | 1 |
| before | yes | 0 |
| after | no | 1 |
| after | yes | 1+ |

The first two constitute at most once and the second two constitute at least once. With that being said, once the retry policy has been exhausted we still need to ack the message and move on even in the case of at least once.